gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/infra/domain_service/subscription/SubscriptionSpecDomainServiceImpl.java

```diff
@@ -0,0 +1,113 @@
+import io.gravitee.apim.core.subscription.model.SubscriptionEntity;
+import io.gravitee.apim.core.subscription.model.crd.SubscriptionSpec;
+import io.gravitee.apim.infra.adapter.SubscriptionAdapter;
+import io.gravitee.rest.api.service.SubscriptionService;
+import io.gravitee.rest.api.service.common.ExecutionContext;
+import io.gravitee.rest.api.service.exceptions.SubscriptionNotFoundException;
+import java.time.ZonedDateTime;
+import java.util.Objects;
+import java.util.Optional;
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.stereotype.Service;
+
+/**
+ * @author Antoine CORDIER (antoine.cordier at graviteesource.com)
+ * @author GraviteeSource Team
+ */
+@Slf4j
+@Service
+@RequiredArgsConstructor
+public class SubscriptionSpecDomainServiceImpl implements SubscriptionSpecDomainService {
+
+    private static final String ACCEPT_REASON = "Kubernetes subscriptions are always auto accepted";
+
+    private final SubscriptionService subscriptionService;
+
+    private final SubscriptionAdapter adapter;
+
+    private final AcceptSubscriptionDomainService acceptService;
+
+    private final CloseSubscriptionDomainService closeSubscriptionDomainService;
+
+    @Override
+    public SubscriptionEntity createOrUpdate(AuditInfo auditInfo, SubscriptionSpec spec) {
+        return find(spec.getId()).map((existing -> update(auditInfo, existing, spec))).orElseGet(() -> create(auditInfo, spec));
+    }
+
+    @Override
+    public void delete(AuditInfo auditInfo, String id) {
+        closeSubscriptionDomainService.closeSubscription(id, auditInfo);
+    }
+
+    private SubscriptionEntity create(AuditInfo auditInfo, SubscriptionSpec spec) {
+        var entity = adapter.fromSpec(spec);
+
+        var then = ZonedDateTime.now();
+
+        entity.setEnvironmentId(auditInfo.environmentId());
+        entity.setSubscribedBy(auditInfo.actor().userId());
+        entity.setProcessedBy(auditInfo.actor().userId());
+        entity.setCreatedAt(then);
+        entity.setProcessedAt(then);
+        entity.setStartingAt(then);
+        entity.setReasonMessage(ACCEPT_REASON);
+        entity.setEndingAt(spec.getEndingAt());
+
+        var subscription = subscriptionService.create(toExecutionContext(auditInfo), adapter.fromCoreForCreate(entity), null, spec.getId());
+
+        if (entity.getEndingAt() != null) {
+            subscriptionService.update(toExecutionContext(auditInfo), adapter.fromCoreForUpdate(entity));
+        }
+
+        log.debug("Auto accepting subscription [{}]", spec.getId());
+        return acceptService.autoAccept(subscription.getId(), ZonedDateTime.now(), spec.getEndingAt(), ACCEPT_REASON, "", auditInfo);
+    }
+
+    private SubscriptionEntity update(AuditInfo auditInfo, SubscriptionEntity existing, SubscriptionSpec spec) {
+        var update = existing.toBuilder().build();
+
+        if (!Objects.equals(spec.getEndingAt(), existing.getEndingAt())) {
```

Comment:
Do we already validate the end date somewhere? Or is it a part of the DryRun?

Reply:
Right now we did not really need a dry run for validating subs, so the format is enforced using a date-time open API data type.